mp/src/game/shared/momentum/util/engine_patch.cpp

```diff
@@ -0,0 +1,243 @@
+// Functions used to find patterns of bytes in the engine's memory to hook or patch
+//------------------------------------------------------------------------------------
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include "Windows.h"
+#include "Psapi.h"
+#pragma comment(lib, "psapi.lib")
+#elif defined (POSIX)
+#include "util/os_utils.h"
+#include "sys/mman.h"
+#endif
+
+#include "engine_patch.h"
+#include "mom_shareddefs.h"
+#include "tier0/platform.h"
+
+void* CEngineBinary::moduleBase = nullptr;
+size_t CEngineBinary::moduleSize;
+
+CEngineBinary::CEngineBinary() : CAutoGameSystem("CEngineBinary")
+{
+}
+
+// Get the engine's base address and size
+bool CEngineBinary::Init()
+{
+#ifdef _WIN32
+    HMODULE handle = GetModuleHandleA("engine.dll");
+    if (!handle)
+        return false;
+
+    MODULEINFO info;
+    GetModuleInformation(GetCurrentProcess(), handle, &info, sizeof(info));
+
+    moduleBase = info.lpBaseOfDll;
+    moduleSize = info.SizeOfImage;
+#else //POSIX
+    if (GetModuleInformation(ENGINE_DLL_NAME, &moduleBase, &moduleSize))
+        return false;
+#endif //WIN32
+
+    return true;
+}
+
+void CEngineBinary::PostInit()
+{
+    CEnginePatch::ApplyAll();
+}
+
+inline bool CEngineBinary::DataCompare(const unsigned char* data, const unsigned char* pattern, const char* mask)
+{
+    for (; *mask != 0; ++data, ++pattern, ++mask)
+        if (*mask == 'x' && *data != *pattern)
+            return false;
+
+    return (*mask == 0);
+}
+
+//---------------------------------------------------------------------------------------------------------
+// Finds a pattern of bytes in the engine memory given a signature and a mask
+// Returns the address of the first (and hopefully only) match with an optional offset, otherwise nullptr
+//---------------------------------------------------------------------------------------------------------
+void* CEngineBinary::FindPattern(const unsigned char* pattern, const char* mask, size_t offset)
+{
+    auto maskLength = strlen(mask);
+    for (size_t i = 0; i <= moduleSize - maskLength; ++i)
+    {
+        auto addr = reinterpret_cast<const unsigned char*>(moduleBase) + i;
+        if (DataCompare(addr, pattern, mask))
+            return const_cast<void*>(reinterpret_cast<const void*>(addr + offset));
+    }
+
+    return nullptr;
+}
+
+CEngineBinary g_EngineBinary;
+
+// Engine Patch format:
+//
+// Patch name
+// Memory signature
+// Signature mask
+// Patch offset
+// Immediate or referenced variable
+// Patch bytes (int/float/char*)
+#ifdef _WIN32
+CEnginePatch g_EnginePatches[] =
+{
+    // Prevent the culling of skyboxes at high FOVs
+    // https://github.com/VSES/SourceEngine2007/blob/master/se2007/engine/gl_warp.cpp#L315
+    // TODO: Use a value derived from FOV instead
+    {
+        "SkyboxCulling",
+        (unsigned char*)"\xF3\x0F\x59\x15\x00\x00\x00\x00\xF3\x0F\x58\xC1\xF3\x0F\x10\x0D",
+        "xxxx????xxxxxxxx",
+        16,
+        false,
```

Comment:
I think using an enum here would be clearer than a bool, hard to remember what true/false corresponds to at a glance.

Reply:
Absolutely